src/core/solvers/ipopt/ipopt-iface.cpp

```diff
@@ -0,0 +1,665 @@
+    for (std::size_t i = 1; i <= ndxi; i++) {
+      nonzero += i;
+    }
+    nnz_h_lag += nonzero;
+  }
+
+  // use the C style indexing (0-based)
+  index_style = Ipopt::TNLP::C_STYLE;
+
+  return true;
+}
+
+bool IpoptInterface::get_bounds_info(Ipopt::Index n, Ipopt::Number *x_l, Ipopt::Number *x_u, Ipopt::Index m,
+                                     Ipopt::Number *g_l, Ipopt::Number *g_u) {
+  assert_pretty(n == nvar_, "Inconsistent number of decision variables");
+  assert_pretty(m == nconst_, "Inconsistent number of constraints");
+
+  // Adding bounds
+  const std::vector<boost::shared_ptr<ActionModelAbstract> > &models = problem_->get_runningModels();
+  std::size_t ixu = 0;
+  for (std::size_t t = 0; t < T_; t++) {
+    // Running state bounds
+    const std::size_t ndxi = models[t]->get_state()->get_ndx();
+    const std::size_t nui = models[t]->get_nu();
+
+    for (std::size_t j = 0; j < ndxi; j++) {
+      x_l[ixu + j] = -2e19;
+      x_u[ixu + j] = 2e19;
+    }
+
+    // Control bounds
+    if (consider_control_bounds_) {
+      for (std::size_t j = 0; j < nui; j++) {
+        x_l[ixu + ndxi + j] = models[t]->get_has_control_limits() ? models[t]->get_u_lb()(j) : -2e19;
+        x_u[ixu + ndxi + j] = models[t]->get_has_control_limits() ? models[t]->get_u_ub()(j) : 2e19;
+      }
+    } else {
+      for (std::size_t j = 0; j < nui; j++) {
+        x_l[ixu + ndxi + j] = -2e19;
+        x_u[ixu + ndxi + j] = 2e19;
+      }
+    }
+
+    ixu += ndxi + nui;
+  }
+
+  {
+    const boost::shared_ptr<ActionModelAbstract> &model = problem_->get_terminalModel();
+    const std::size_t ndxi = model->get_state()->get_ndx();
+
+    // Final state bounds
+    for (std::size_t j = 0; j < ndxi; j++) {
+      x_l[ixu + j] = -2e19;
+      x_u[ixu + j] = 2e19;
+    }
+  }
+
+  // Dynamics & Initial conditions (all equal to zero)
+  for (Ipopt::Index i = 0; i < nconst_; i++) {
+    g_l[i] = 0;
+    g_u[i] = 0;
+  }
+
+  return true;
+}
+
+bool IpoptInterface::get_starting_point(Ipopt::Index n, bool init_x, Ipopt::Number *x, bool init_z, Ipopt::Number *z_L,
+                                        Ipopt::Number *z_U, Ipopt::Index m, bool init_lambda, Ipopt::Number *lambda) {
+  assert_pretty(init_x == true, "Make sure to provide initial value for primal variables");
+  assert_pretty(init_z == false, "Cannot provide initial value for bound multipliers");
+  assert_pretty(init_lambda == false, "Cannot provide initial value for constraint multipliers");
+
+  // initialize to the given starting point
+  // State variables are always at 0 since they represent increments from the given initial point
+  const std::vector<boost::shared_ptr<ActionModelAbstract> > &models = problem_->get_runningModels();
+  std::size_t ixu = 0;
+  for (std::size_t t = 0; t < T_; t++) {
+    const std::size_t ndxi = models[t]->get_state()->get_ndx();
+    const std::size_t nui = models[t]->get_nu();
+
+    for (std::size_t j = 0; j < ndxi; j++) {
+      x[ixu + j] = 0;
```

Comment:
It seems to me that we should use the deltas xs. You should compute as xs_[i] (-) 0 = diff(0, xs_[i])

Reply:
I've implemented the following: $\mathbf{x}(0) \ominus (\mathbf{x}^0_k \oplus \mathbf{\Delta x}_k) = \mathbf{0}$